vhost-device-input/src/main.rs

```diff
@@ -0,0 +1,198 @@
+//
+// Copyright 2023 Linaro Ltd. All Rights Reserved.
+// Leo Yan <leo.yan@linaro.org>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+mod input;
+mod vhu_input;
+
+use log::error;
+use std::os::fd::AsRawFd;
+use std::path::PathBuf;
+use std::process::exit;
+use std::sync::{Arc, RwLock};
+use std::thread::{self, JoinHandle};
+
+use clap::Parser;
+use thiserror::Error as ThisError;
+use vhost_user_backend::VhostUserDaemon;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+use vhu_input::VuInputBackend;
+use vmm_sys_util::epoll::EventSet;
+
+#[derive(Debug, ThisError)]
+/// Errors related to vhost-device-input daemon.
+pub(crate) enum Error {
+    #[error("Event device file doesn't exists or can't be accessed")]
+    AccessEventDeviceFile,
+    #[error("Threads can't be joined")]
+    FailedJoiningThreads,
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(std::io::Error),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Could not register input event into vring epoll")]
+    CouldNotRegisterInputEvent,
+    #[error("Fatal error: {0}")]
+    ServeFailed(vhost_user_backend::Error),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Clone, Parser, Debug, PartialEq)]
+#[clap(author, version, about, long_about = None)]
+struct InputArgs {
+    // Location of vhost-user Unix domain socket.
+    #[clap(short, long)]
+    socket_path: PathBuf,
+
+    // Path for reading input events.
+    #[clap(short = 'e', long)]
+    event_list: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct VuInputConfig {
+    pub socket_path: PathBuf,
+    pub events: Vec<PathBuf>,
+}
+
+impl TryFrom<InputArgs> for VuInputConfig {
+    type Error = Error;
+
+    fn try_from(args: InputArgs) -> Result<Self> {
+        let mut events: Vec<PathBuf> = Vec::new();
+
+        for ev_str in args.event_list.trim().split(',') {
+            let ev_path = PathBuf::from(ev_str.to_string());
+            events.push(ev_path);
+        }
+
+        Ok(VuInputConfig {
+            socket_path: args.socket_path,
+            events,
+        })
+    }
+}
+
+pub(crate) fn start_backend(config: VuInputConfig) -> Result<()> {
+    let mut handles = Vec::new();
+
+    for (i, event) in config.events.iter().cloned().enumerate() {
+        let socket = format!("{}{}", config.socket_path.display(), i);
+
+        let handle: JoinHandle<Result<()>> = thread::spawn(move || loop {
+            let ev_dev = match evdev::Device::open(&event) {
+                Ok(dev) => dev,
+                Err(err) => {
+                    error!(
+                        "Failed to open event device {}, error is: {}",
+                        event.display(),
+                        err
+                    );
+                    return Err(Error::AccessEventDeviceFile);
+                }
+            };
```

Comment:
I will change the code to:

```
let ev_dev = D::open(event.clone()).map_err(|_| Error::AccessEventDeviceFile)?;
```

But I found when the open() function returns error, it doesn't print error log.  It's likely because it's in the thread loop.  I confirmed it outputs error log outside the thread loop.